examples/push-repo

```diff
@@ -0,0 +1,194 @@
+    parser.add_argument(
+        "--file-metadata",
+        default={},
+        type=json.loads,
+        help="Additional data sent with the file in JSON string format,"
+        ' e.g., \'{"key": "value"}\'',
+    )
+    parser.add_argument(
+        "--release-date",
+        help="Date on which the content will be made available.",
+    )
+    parser.add_argument(
+        "--bucket", required=True, help="S3 bucket in which to upload file."
+    )
+    parser.add_argument(
+        "--table",
+        required=True,
+        help="DynamoDB table in which to publish file data.",
+    )
+    parser.add_argument(
+        "--aws-access-id",
+        default=None,
+        help="Access ID for Amazon services. If no ID is provided, attempts to"
+        " find it among environment variables and ~/.aws/config file will"
+        " be made",
+    )
+    parser.add_argument(
+        "--aws-access-key",
+        default=None,
+        help="Access key for Amazon services. If no key is provided, attempts"
+        " to find it among environment variables and ~/.aws/config file"
+        " will be made",
+    )
+    parser.add_argument(
+        "--aws-session-token",
+        default=None,
+        help="Session token for Amazon services. If no token is provided,"
+        " attempts to find it among environment variables and"
+        " ~/.aws/config file will be made",
+    )
+    parser.add_argument(
+        "--default-region",
+        default=None,
+        help="Default region for Amamzon services. If no region is provided,"
+        " attempts to find it among environment variables and"
+        " ~/.aws/config file will be made",
+    )
+    parser.add_argument(
+        "--dryrun",
+        action="store_true",
+        help="Don't execute the action, only log what would otherwise be done.",
+    )
+    parser.add_argument(
+        "--debug", action="store_true", help="Include debug logging."
+    )
+
+    p = parser.parse_args()
+
+    if p.debug:
+        logging.getLogger("chexus").setLevel(logging.DEBUG)
+        LOG.setLevel(logging.DEBUG)
+
+    client = Client(
+        access_id=p.aws_access_id,
+        access_key=p.aws_access_key,
+        session_token=p.aws_session_token,
+        default_region=p.default_region,
+    )
+
+    for root_dir, subdir_list, file_list in os.walk(p.local_repo):
+        for file_name in file_list:
+            file_path = os.path.join(root_dir, file_name)
+            upl_item = BucketItem(file_path)
+            upl_item.key = upl_item.checksum
+
+            origin_item = TableItem(
+                object_key=upl_item.checksum,
+                web_uri=cdn_path(upl_item.name, file_path, upl_item.checksum),
+                from_date=p.release_date,
+                metadata=p.file_metadata,
+            )
+
+            repo_item = TableItem(
+                object_key=upl_item.checksum,
+                web_uri=os.path.join(
+                    p.dest_repo, get_dir_name(root_dir), file_name
```

Comment:
I'm not quite getting this part about calling get_dir_name and using it to assemble a URI.

Why can't we use the full path under root_dir without further processing?  I mean, join p.dest_repo and file_path.

The way it is now seems to potentially change the structure of the repo during publish, which would break the repo.  e.g. yum repos are allowed to have subdirs under Packages, and most Pulp/CDN repos do have that, but get_dir_name would translate e.g. "Packages/b" into just "" and end up publishing RPMs to the wrong place.

Reply:
This comment is the reason for my -1, other comments are not as important.

Reply:
> Why can't we use the full path under root_dir without further processing? I mean, join p.dest_repo and file_path.

file_path is the full path of the file on the local file system. Joining p.dest_repo and file_path would result in:
`/content/beta/rhel8/8/x86_64/exodus/iso/home/crungeho/exodus/isorepo/some-rhel-dvd.iso`.

> The way it is now seems to potentially change the structure of the repo during publish, which would break the repo. e.g. yum repos are allowed to have subdirs under Packages, and most Pulp/CDN repos do have that, but get_dir_name would translate e.g. "Packages/b" into just "" and end up publishing RPMs to the wrong place.

Ah, good point. I've corrected this. Example:
```
Would publish the following item to the 'path-to-object' table;
	{
    "from_date": "2020-02-01T10:00:00",
    "metadata": {},
    "object_key": "5976a32244ccd1e5cabf8fef04f7c65eae6802504344a14f08b9539f941189df",
    "web_uri": "/content/beta/rhel8/8/x86_64/exodus/os/Packages/b/bash-4.4.19-7.el8.x86_64.rpm"
}
```